### PR TITLE
Make Push compliant with App Server role in RFC8030

### PR DESCRIPTION
**Pull request by neilj**

### CODE REVIEW

File: spec/jmap/push.mdown [hunk 1]
```diff
@@ -26,25 +31,39 @@ Refer to the Authentication section of this spec for details on how to get the U
 
 A client MAY hold open multiple connections to the event-source, although it SHOULD try to use a single connection for efficiency.
 
-## Web hook
+## Callback URL
 
-The second push mechanism is to register a callback URL to which the JMAP server will make an HTTPS POST request whenever the event occurs. The request MUST have a content type of `application/json` and contain the same UTF-8 JSON encoded object as described above as the body.
+The second push mechanism is to register a callback URL to which the JMAP server will make an HTTPS POST request whenever an event occurs. This is compatible with the Application Server role of [@!RFC8030].
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: What is compatible?  The URL, the thing using the URL, or the resource that the URL refers to?

File: spec/jmap/push.mdown [hunk 2]
```diff
@@ -26,25 +31,39 @@ Refer to the Authentication section of this spec for details on how to get the U
...
 The callback is tied to the access token used to create it. Should the access token expire or be revoked, the callback MUST be removed by the JMAP server. The client MUST re-register the callback after reauthenticating to resume callbacks.
 
 Each session may only have a single callback URL registered. It can be set or retrieved using the following API calls.
 
 ### setPushCallback
 
-To set the web hook, make a call to *setPushCallback*. It takes the following argument:
+To set the callback URL, make a call to *setPushCallback*. It takes the following argument:
 
 - **callback**: `String|null`
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: This isn't sufficient.  See https://w3c.github.io/push-api/#dom-pushsubscription for a description of the properties that are needed to use (most implementations of) RFC 8030.

File: spec/jmap/push.mdown [hunk 3]
```diff
@@ -56,7 +75,7 @@ The response to *setPushCallback* is called *pushCallbackSet*. It has the follow
 
 The following error may be returned instead of the *mailboxesSet* response:
 
-`invalidUrl`: Returned if the URL does not begin with `https://`, or is otherwise syntactically invalid or does not resolve.
+`invalidUrl`: Returned if the URL does not begin with `https://`, or is otherwise syntactically invalid or cannot be used by the server.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: Invalid or unsupported?

File: spec/jmap/push.mdown [hunk 4]
```diff
@@ -26,25 +31,39 @@ Refer to the Authentication section of this spec for details on how to get the U
...
+
+The JMAP server MAY include the following headers in the POST request:
+
+- `Urgency`: One of `very-low`, `low`, `normal`, or `high` as per [@!RFC8030]
+  section 5.3.
+- `Topic`: A token that encodes the set of types included in the *changed*
+  property of this push request. This MUST be restricted to no more than 32
+  characters from the URL and filename-safe Base 64 alphabet ([@!RFC4648]), as
+  per [@!RFC8030] section 5.4. This allows the push server to replace
+  undelivered messages which are entirely superceded by the later message.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: This just repeats what RFC 8030 says.  Are there any semantics that you would like to attach to these (I think that the answer is, and should be, no)?  If there aren't semantics, I don't see why you would mention them.

File: spec/jmap/push.mdown [hunk 5]
```diff
@@ -1,11 +1,16 @@
...
+Push notifications allow clients to efficiently update (almost) instantly to stay in sync with data changes on the server. In JMAP, push notifications occur out-of-band (i.e. not over the same connection as API exchanges), so that they can make use of efficient native push mechanisms on different platforms.
 
 The general model for push is simple and does not send any sensitive data over the push channel, making it suitable for use with less trusted 3rd party intermediaries. The format allows multiple changes to be coalesced into a single push update, and the frequency of pushes to be rate limited by the server. It doesn't matter if some push events are dropped before they reach the client; it will still get all changes next time it syncs.
 
-When something changes on the server, the server pushes a small JSON object to the client with the following property:
+When something changes on the server, the server pushes a JSON object to the client with the following properties:
 
+- **type**: `String`
+  This MUST be the string `"state"`. This is to allow future extension of the
+  push protocol with other event types.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: What are the valid values for this, and what do they each mean?

File: spec/jmap/push.mdown [hunk 6]
```diff
@@ -1,11 +1,16 @@
...
+- **username**: `String`
+  The username of the authenticated user.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: This isn't necessary: a subscription can be created that is bound to a specific context.  (Subscriptions are cheap.)

**Comment by neilj** [comment]
@martinthomson If you have a moment to review the revised pull request, would be much appreciated. I believe all your previous comments have been addressed.

File: spec/jmap/push.mdown [hunk 7]
```diff
@@ -1,69 +1,94 @@
...
-MAY choose not to do so for policy reasons. If the final response code from the server is `2xx`, the callback is considered a success. If the response code is `503` (Service Unavailable), the JMAP server MAY try again later (but may also just drop the event). If the response code is `429` (Too Many Requests) the JMAP server SHOULD attempt to reduce the frequency of pushes to that URL. Any other response code MUST be considered a **permanent failure** and the callback should be deregistered (not tried again even for future events unless explicitly re-registered by the client).
+The POST request MUST have a content type of `application/json` and contain the utf-8 JSON encoded *StateChange* object as the body. The request MUST have a `TTL` header, and MAY have `Urgency` and/or `Topic` headers, as specified in section 5 of [@!RFC8030].
 
-The URL set by the client MUST use the HTTPS protocol and SHOULD encode within it a unique token that can be verified by the server to know that the request comes from the JMAP server the authenticated client connected to.
+The JMAP server MUST follow at least one redirect, and MAY follow more.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: This seems overly prescriptive.  Are you talking about profiling HTTP?

File: spec/jmap/push.mdown [hunk 8]
```diff
@@ -1,69 +1,94 @@
...
+A push subscription is a message delivery context established between the client and a push service. A **PushSubscription** object has the following properties:
 
-## Web hook
+- **url**: `String`
+  An absolute URL where the JMAP server will POST the data for the push message.
+  This MUST begin with `https://`.
+- **expires**: `Date|null`
+  The time this push subscription expires. If specified, the JMAP server MUST NOT make further requests to this resource after this time. It MAY automatically remove the push subscription at or after this time.
+- **keys**: `Object|null`
+  Clients generated encryption keys. If supplied the server MUST use them as specified in <https://tools.ietf.org/html/draft-ietf-webpush-encryption-09> to encrypt all data sent to the push subscription. The object MUST have the following properties:
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: I don't like that encryption is optional, but realize that it's only extra trouble if you are talking directly (as opposed to via a push service).  Can you describe exactly what conditions are necessary for the keys to be left unset?

File: spec/jmap/push.mdown [hunk 9]
```diff
@@ -1,69 +1,94 @@
...
+- **pushSubscription**: `PushSubscription|null`
+  The PushSubscription object the JMAP server is currently posting push events to, or `null` if none.
+
+## Event Source
+
+Clients that can hold open TCP connections can connect directly to the JMAP server to receive push notifications via a `text/event-stream` resource, as described in <http://www.w3.org/TR/eventsource/>. This is a long running HTTP request down which the server can push data.
+
+When a change occurs in the data on the server, it pushes an event called **state** to any connected clients, with the *StateChange* object as the data.
+
+The server SHOULD also set a new `Last-Event-Id` that encodes the entire server state visible to the user. When a new connection is made to the event-source endpoint, the server can then work out whether the client has missed some changes which it should send immediately.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: I don't have enough context to understand how to use this Last-Event-Id.  Do you have a reference to other parts of the document, or can you expand the description. (Where does the server set this? Does the client set something?)

File: spec/jmap/push.mdown [hunk 10]
```diff
@@ -1,69 +1,94 @@
...
+
+The server MUST also send an event called **ping** with an empty object as the data if a maximum of 5 minutes has elapsed since the previous event. This MUST NOT set a new `Last-Event-Id`.
```
Comment (martinthomson) [inline spec/jmap/push.mdown]: 5 minutes is pretty arbitrary.  Why that specific number?  Can I use my knowledge about the network path to choose a longer (and therefore more efficient) value?

Comment (neilj) [reply spec/jmap/push.mdown]: Yeh, it's arbitrary. The trouble is the client needs to know what value the server is using to know if it should have seen a ping by now! The client could set this in a header when it connects to the server though to request a particular ping time; that's probably better. The main downside is that you can't set custom headers if you are using the browsers' native EventSource object (web clients can still use XHR to handle it though).

File: spec/jmap/push.mdown (hunk 9 above)
Comment (neilj) [reply spec/jmap/push.mdown]: The concept is a standard part of [Server Sent Events](https://html.spec.whatwg.org/multipage/server-sent-events.html), which is why I didn't describe it in too much detail, but I can add more.

Basically, the server can send an id with an event in the text/event-stream. If the client loses its connection, when it reconnects it sends the server the last id it saw as a Last-Event-ID header in the HTTP request. The server can then detect if the client missed any events while it was disconnected and send them immediately if so.

File: spec/jmap/push.mdown (hunk 7 above)
Comment (neilj) [reply spec/jmap/push.mdown]: No, that whole line is unnecessary; I'll cut it.

File: spec/jmap/push.mdown (hunk 8 above)
Comment (neilj) [reply spec/jmap/push.mdown]: Well, basically if the push server is owned by the same entity as the JMAP server, you don't gain anything via the encryption and it just adds overhead. If it's going via a 3rd party then the client should use encryption. Something like that, wordsmithed a bit?

**Pull request merged**